flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientSavepointTriggerTest.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.flink.runtime.rest.messages.job.savepoints.SavepointTriggerHeaders;
+import org.apache.flink.runtime.rest.messages.job.savepoints.SavepointTriggerMessageParameters;
+import org.apache.flink.runtime.rest.messages.job.savepoints.SavepointTriggerRequestBody;
+import org.apache.flink.runtime.rpc.RpcUtils;
+import org.apache.flink.runtime.util.ExecutorThreadFactory;
+import org.apache.flink.runtime.webmonitor.TestingDispatcherGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.util.ConfigurationException;
+import org.apache.flink.util.SerializedThrowable;
+import org.apache.flink.util.TestLogger;
+import org.apache.flink.util.function.FunctionWithException;
+
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandler;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests for the {@link RestClusterClient} for operations that trigger savepoints.
+ *
+ * <p>These tests verify that the client uses the appropriate headers for each
+ * request, properly constructs the request bodies/parameters and processes the responses correctly.
+ */
+public class RestClusterClientSavepointTriggerTest extends TestLogger {
+
+	private static final DispatcherGateway mockRestfulGateway = new TestingDispatcherGateway.Builder().build();
+
+	private static final GatewayRetriever<DispatcherGateway> mockGatewayRetriever = () -> CompletableFuture.completedFuture(mockRestfulGateway);
+
+	private static RestServerEndpointConfiguration restServerEndpointConfiguration;
+
+	private static ExecutorService executor;
+
+	private static final Configuration restConfig;
+
+	static {
+		final Configuration config = new Configuration();
+		config.setString(JobManagerOptions.ADDRESS, "localhost");
+		config.setInteger(RestOptions.RETRY_MAX_ATTEMPTS, 10);
+		config.setLong(RestOptions.RETRY_DELAY, 0);
+		config.setInteger(RestOptions.PORT, 0);
+
+		restConfig = config;
+	}
+
+	@BeforeClass
+	public static void setUp() throws ConfigurationException {
+		restServerEndpointConfiguration = RestServerEndpointConfiguration.fromConfiguration(restConfig);
```

Comment:
`restServerEndpointConfiguration` could be as well `static final` and initialized in the static initializer.

Reply:
One could, but `RestServerEndpointConfiguration.fromConfiguration` may throw exception and doing so in the static blocks leads to misleading stacktraces and odd behaviors.

If this call were to fail the first time the test is run you'll get a ExceptionOnInitializationError (i.e. the classes cannot be loaded), but the JVM doesn't try to load the class again resulting in, iirc, ClassNotFoundExceptions or something.

Reply:
ok